{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: biom-format in /home/jelena/EMAN2/envs/microbiome_toolbox/lib/python3.6/site-packages (2.1.10)\n",
      "Requirement already satisfied: h5py in /home/jelena/EMAN2/envs/microbiome_toolbox/lib/python3.6/site-packages (from biom-format) (3.1.0)\n",
      "Requirement already satisfied: pandas>=0.20.0 in /home/jelena/EMAN2/envs/microbiome_toolbox/lib/python3.6/site-packages (from biom-format) (1.1.3)\n",
      "Requirement already satisfied: click in /home/jelena/EMAN2/envs/microbiome_toolbox/lib/python3.6/site-packages (from biom-format) (7.1.2)\n",
      "Requirement already satisfied: six>=1.10.0 in /home/jelena/EMAN2/envs/microbiome_toolbox/lib/python3.6/site-packages (from biom-format) (1.15.0)\n",
      "Requirement already satisfied: cython in /home/jelena/EMAN2/envs/microbiome_toolbox/lib/python3.6/site-packages (from biom-format) (0.29.24)\n",
      "Requirement already satisfied: scipy>=1.3.1 in /home/jelena/EMAN2/envs/microbiome_toolbox/lib/python3.6/site-packages (from biom-format) (1.5.2)\n",
      "Requirement already satisfied: numpy>=1.9.2 in /home/jelena/EMAN2/envs/microbiome_toolbox/lib/python3.6/site-packages (from biom-format) (1.19.1)\n",
      "Requirement already satisfied: future>=0.16.0 in /home/jelena/EMAN2/envs/microbiome_toolbox/lib/python3.6/site-packages (from biom-format) (0.18.2)\n",
      "Requirement already satisfied: cached-property; python_version < \"3.8\" in /home/jelena/EMAN2/envs/microbiome_toolbox/lib/python3.6/site-packages (from h5py->biom-format) (1.5.2)\n",
      "Requirement already satisfied: python-dateutil>=2.7.3 in /home/jelena/EMAN2/envs/microbiome_toolbox/lib/python3.6/site-packages (from pandas>=0.20.0->biom-format) (2.8.1)\n",
      "Requirement already satisfied: pytz>=2017.2 in /home/jelena/EMAN2/envs/microbiome_toolbox/lib/python3.6/site-packages (from pandas>=0.20.0->biom-format) (2020.1)\n"
     ]
    }
   ],
   "source": [
    "!pip install biom-format"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 326,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting kaleido\n",
      "  Downloading kaleido-0.2.1-py2.py3-none-manylinux1_x86_64.whl (79.9 MB)\n",
      "\u001b[K     |████████████████████████████████| 79.9 MB 259 kB/s eta 0:00:01\n",
      "\u001b[?25hInstalling collected packages: kaleido\n",
      "Successfully installed kaleido-0.2.1\n"
     ]
    }
   ],
   "source": [
    "!pip install -U kaleido"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {},
   "outputs": [],
   "source": [
    "from biom import load_table\n",
    "input_biom = \"../Human_Subramanian/INPUT_FILES/Subramanian_et_al_otu_table/Subramanian_et_al_otu_table.biom\"\n",
    "table = load_table(input_biom)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(1897, 79598)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sampleID</th>\n",
       "      <th>304309</th>\n",
       "      <th>16076</th>\n",
       "      <th>373909</th>\n",
       "      <th>35786</th>\n",
       "      <th>591907</th>\n",
       "      <th>410908</th>\n",
       "      <th>117676</th>\n",
       "      <th>326792</th>\n",
       "      <th>189116</th>\n",
       "      <th>...</th>\n",
       "      <th>GGS.ReferenceOTU193</th>\n",
       "      <th>GGS.ReferenceOTU191</th>\n",
       "      <th>GGS.ReferenceOTU190</th>\n",
       "      <th>GGS.ReferenceOTU197</th>\n",
       "      <th>GGS.ReferenceOTU195</th>\n",
       "      <th>GGS.ReferenceOTU199</th>\n",
       "      <th>GGS.ReferenceOTU294</th>\n",
       "      <th>GGS.ReferenceOTU66</th>\n",
       "      <th>GGS.ReferenceOTU61</th>\n",
       "      <th>GGS.ReferenceOTU229</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Bgsng7114.m22</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>507</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Bgsng8064.m14</td>\n",
       "      <td>11</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Bgsng7082.m24</td>\n",
       "      <td>0</td>\n",
       "      <td>1369</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>81</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Bgsng7192.m22</td>\n",
       "      <td>0</td>\n",
       "      <td>47</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>39</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Bgsng7082.m19</td>\n",
       "      <td>0</td>\n",
       "      <td>48</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 79598 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        sampleID  304309  16076  373909  35786  591907  410908  117676  \\\n",
       "0  Bgsng7114.m22       1      0       0      0       0       0       0   \n",
       "1  Bgsng8064.m14      11      0       0      0       0       0       0   \n",
       "2  Bgsng7082.m24       0   1369       0      0       0       0       0   \n",
       "3  Bgsng7192.m22       0     47       0      0       0       0       0   \n",
       "4  Bgsng7082.m19       0     48       0      0       0       0       0   \n",
       "\n",
       "   326792  189116  ...  GGS.ReferenceOTU193  GGS.ReferenceOTU191  \\\n",
       "0     507       0  ...                    0                    0   \n",
       "1       1       0  ...                    0                    0   \n",
       "2      81       0  ...                    0                    0   \n",
       "3      39       0  ...                    0                    0   \n",
       "4       4       0  ...                    0                    0   \n",
       "\n",
       "   GGS.ReferenceOTU190  GGS.ReferenceOTU197  GGS.ReferenceOTU195  \\\n",
       "0                    0                    0                    0   \n",
       "1                    0                    0                    0   \n",
       "2                    0                    0                    0   \n",
       "3                    0                    0                    0   \n",
       "4                    0                    0                    0   \n",
       "\n",
       "   GGS.ReferenceOTU199  GGS.ReferenceOTU294  GGS.ReferenceOTU66  \\\n",
       "0                    0                    0                   0   \n",
       "1                    0                    0                   0   \n",
       "2                    0                    0                   0   \n",
       "3                    0                    0                   0   \n",
       "4                    0                    0                   0   \n",
       "\n",
       "   GGS.ReferenceOTU61  GGS.ReferenceOTU229  \n",
       "0                   0                    0  \n",
       "1                   0                    0  \n",
       "2                   0                    0  \n",
       "3                   0                    0  \n",
       "4                   0                    0  \n",
       "\n",
       "[5 rows x 79598 columns]"
      ]
     },
     "execution_count": 90,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_abundances = table.to_dataframe().T.rename_axis('sampleID').astype(np.int32).reset_index()\n",
    "print(df_abundances.shape)\n",
    "df_abundances.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0        1\n",
       "1       11\n",
       "2        0\n",
       "3        0\n",
       "4        0\n",
       "        ..\n",
       "1892     0\n",
       "1893     0\n",
       "1894     0\n",
       "1895     0\n",
       "1896     0\n",
       "Name: 304309, Length: 1897, dtype: Sparse[int32, 0]"
      ]
     },
     "execution_count": 91,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_abundances[\"304309\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {},
   "outputs": [],
   "source": [
    "file_name = \"../Human_Subramanian/INPUT_FILES/Subramanian_et_al_mapping_file.txt\"\n",
    "df_metadata = pd.read_csv(file_name, sep=None, engine='python')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['#SampleID', 'BarcodeSequence', 'LinkerPrimerSequence', 'PersonID',\n",
       "       'FamilyID', 'ENA.SampleID', 'ENA.LibraryName', 'Health_Analysis_Groups',\n",
       "       'tax_id', 'scientific_name', 'common_name', 'anonymized_name',\n",
       "       'sample_title', 'sample_description', 'investigation type',\n",
       "       'project name', 'target gene', 'sequencing method', 'collection date',\n",
       "       'host subject id', 'age_in_months', 'sex', 'environmental package',\n",
       "       'geographic location_latitude', 'geographic location_longitude',\n",
       "       'geographic location_country', 'environment_biome',\n",
       "       'environment_feature', 'environment_material', 'Description'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 93,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_metadata.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {},
   "outputs": [],
   "source": [
    "# TODO: handle group and reference_group"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_metadata.age_in_months *= 30"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# id_columns = [\"FamilyID\", \"ENA.SampleID\", \"tax_id\"]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_metadata = df_metadata.rename({\"#SampleID\": \"sampleID\", \n",
    "               \"PersonID\": \"subjectID\",\n",
    "               \"age_in_months\": \"age_at_collection\",}, axis=1)   \n",
    "\n",
    "other_names = {}\n",
    "for col1 in set(df_metadata.columns).difference(set(MicrobiomeDataset.specific_columns)):\n",
    "    other_names[col1] = f\"meta_{col1}\"\n",
    "    \n",
    "df_metadata = df_metadata.rename(other_names, axis=1)        \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "meta_LinkerPrimerSequence has only 1 unique value [nan]\n",
      "meta_tax_id has only 1 unique value [408170]\n",
      "meta_scientific_name has only 1 unique value ['human gut metagenome']\n",
      "meta_common_name has only 1 unique value ['human gut metagenome']\n",
      "meta_sample_description has only 1 unique value ['human gut metagenome']\n",
      "meta_investigation type has only 1 unique value ['mimarks-survey']\n",
      "meta_project name has only 1 unique value ['Subramanian maturity']\n",
      "meta_target gene has only 1 unique value ['V4']\n",
      "meta_sequencing method has only 1 unique value ['Illumina MiSeq']\n",
      "meta_collection date has only 1 unique value ['None']\n",
      "meta_environmental package has only 1 unique value ['human-gut']\n",
      "meta_environment_biome has only 1 unique value ['ENVO:human-associated habitat']\n",
      "meta_environment_feature has only 1 unique value ['ENVO:human-associated habitat']\n",
      "meta_environment_material has only 1 unique value ['ENVO:feces']\n"
     ]
    }
   ],
   "source": [
    "columns_to_remove = [\"ENA.SampleID\", \n",
    "                     \"anonymized_name\",\n",
    "                     \"sample_title\",\n",
    "                     \"host subject id\",\n",
    "                     \"Description\"]\n",
    "for col in df_metadata.columns:\n",
    "    if len(df_metadata[col].unique()) == 1:\n",
    "        print(col, \"has only 1 unique value\", df_metadata[col].unique())\n",
    "        df_metadata.drop(col, inplace=True, axis=1)\n",
    "    elif col in columns_to_remove:\n",
    "        df_metadata.drop(col, inplace=True, axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['sampleID', 'meta_BarcodeSequence', 'subjectID', 'meta_FamilyID',\n",
       "       'meta_ENA.SampleID', 'meta_ENA.LibraryName',\n",
       "       'meta_Health_Analysis_Groups', 'meta_anonymized_name',\n",
       "       'meta_sample_title', 'meta_host subject id', 'age_at_collection',\n",
       "       'meta_sex', 'meta_geographic location_latitude',\n",
       "       'meta_geographic location_longitude',\n",
       "       'meta_geographic location_country', 'meta_Description'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 99,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_metadata.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(1897, 16)"
      ]
     },
     "execution_count": 100,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_metadata.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.merge(df_abundances, df_metadata, on='sampleID')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(1894, 79613)"
      ]
     },
     "execution_count": 102,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sampleID</th>\n",
       "      <th>304309</th>\n",
       "      <th>16076</th>\n",
       "      <th>373909</th>\n",
       "      <th>35786</th>\n",
       "      <th>591907</th>\n",
       "      <th>410908</th>\n",
       "      <th>117676</th>\n",
       "      <th>326792</th>\n",
       "      <th>189116</th>\n",
       "      <th>...</th>\n",
       "      <th>meta_Health_Analysis_Groups</th>\n",
       "      <th>meta_anonymized_name</th>\n",
       "      <th>meta_sample_title</th>\n",
       "      <th>meta_host subject id</th>\n",
       "      <th>age_at_collection</th>\n",
       "      <th>meta_sex</th>\n",
       "      <th>meta_geographic location_latitude</th>\n",
       "      <th>meta_geographic location_longitude</th>\n",
       "      <th>meta_geographic location_country</th>\n",
       "      <th>meta_Description</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Bgsng7114.m22</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>507</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>Healthy Singletons</td>\n",
       "      <td>Bgsng7114</td>\n",
       "      <td>Bgsng7114.m22</td>\n",
       "      <td>Bgsng7114</td>\n",
       "      <td>630.0</td>\n",
       "      <td>male</td>\n",
       "      <td>23.933</td>\n",
       "      <td>88.983</td>\n",
       "      <td>Bangladesh</td>\n",
       "      <td>Bgsng7114.m22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Bgsng8064.m14</td>\n",
       "      <td>11</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>Healthy Singletons</td>\n",
       "      <td>Bgsng8064</td>\n",
       "      <td>Bgsng8064.m14</td>\n",
       "      <td>Bgsng8064</td>\n",
       "      <td>393.0</td>\n",
       "      <td>male</td>\n",
       "      <td>23.933</td>\n",
       "      <td>88.983</td>\n",
       "      <td>Bangladesh</td>\n",
       "      <td>Bgsng8064.m14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Bgsng7082.m24</td>\n",
       "      <td>0</td>\n",
       "      <td>1369</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>81</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>Healthy Singletons</td>\n",
       "      <td>Bgsng7082</td>\n",
       "      <td>Bgsng7082.m24</td>\n",
       "      <td>Bgsng7082</td>\n",
       "      <td>714.0</td>\n",
       "      <td>female</td>\n",
       "      <td>23.933</td>\n",
       "      <td>88.983</td>\n",
       "      <td>Bangladesh</td>\n",
       "      <td>Bgsng7082.m24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Bgsng7192.m22</td>\n",
       "      <td>0</td>\n",
       "      <td>47</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>39</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>Healthy Singletons</td>\n",
       "      <td>Bgsng7192</td>\n",
       "      <td>Bgsng7192.m22</td>\n",
       "      <td>Bgsng7192</td>\n",
       "      <td>636.0</td>\n",
       "      <td>female</td>\n",
       "      <td>23.933</td>\n",
       "      <td>88.983</td>\n",
       "      <td>Bangladesh</td>\n",
       "      <td>Bgsng7192.m22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Bgsng7082.m19</td>\n",
       "      <td>0</td>\n",
       "      <td>48</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>Healthy Singletons and MAM cross-sectional ana...</td>\n",
       "      <td>Bgsng7082</td>\n",
       "      <td>Bgsng7082.m19</td>\n",
       "      <td>Bgsng7082</td>\n",
       "      <td>543.0</td>\n",
       "      <td>female</td>\n",
       "      <td>23.933</td>\n",
       "      <td>88.983</td>\n",
       "      <td>Bangladesh</td>\n",
       "      <td>Bgsng7082.m19</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 79613 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        sampleID  304309  16076  373909  35786  591907  410908  117676  \\\n",
       "0  Bgsng7114.m22       1      0       0      0       0       0       0   \n",
       "1  Bgsng8064.m14      11      0       0      0       0       0       0   \n",
       "2  Bgsng7082.m24       0   1369       0      0       0       0       0   \n",
       "3  Bgsng7192.m22       0     47       0      0       0       0       0   \n",
       "4  Bgsng7082.m19       0     48       0      0       0       0       0   \n",
       "\n",
       "   326792  189116  ...                        meta_Health_Analysis_Groups  \\\n",
       "0     507       0  ...                                 Healthy Singletons   \n",
       "1       1       0  ...                                 Healthy Singletons   \n",
       "2      81       0  ...                                 Healthy Singletons   \n",
       "3      39       0  ...                                 Healthy Singletons   \n",
       "4       4       0  ...  Healthy Singletons and MAM cross-sectional ana...   \n",
       "\n",
       "   meta_anonymized_name  meta_sample_title  meta_host subject id  \\\n",
       "0             Bgsng7114      Bgsng7114.m22             Bgsng7114   \n",
       "1             Bgsng8064      Bgsng8064.m14             Bgsng8064   \n",
       "2             Bgsng7082      Bgsng7082.m24             Bgsng7082   \n",
       "3             Bgsng7192      Bgsng7192.m22             Bgsng7192   \n",
       "4             Bgsng7082      Bgsng7082.m19             Bgsng7082   \n",
       "\n",
       "   age_at_collection  meta_sex  meta_geographic location_latitude  \\\n",
       "0              630.0      male                             23.933   \n",
       "1              393.0      male                             23.933   \n",
       "2              714.0    female                             23.933   \n",
       "3              636.0    female                             23.933   \n",
       "4              543.0    female                             23.933   \n",
       "\n",
       "   meta_geographic location_longitude  meta_geographic location_country  \\\n",
       "0                              88.983                        Bangladesh   \n",
       "1                              88.983                        Bangladesh   \n",
       "2                              88.983                        Bangladesh   \n",
       "3                              88.983                        Bangladesh   \n",
       "4                              88.983                        Bangladesh   \n",
       "\n",
       "   meta_Description  \n",
       "0     Bgsng7114.m22  \n",
       "1     Bgsng8064.m14  \n",
       "2     Bgsng7082.m24  \n",
       "3     Bgsng7192.m22  \n",
       "4     Bgsng7082.m19  \n",
       "\n",
       "[5 rows x 79613 columns]"
      ]
     },
     "execution_count": 103,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "sampleID                                        object\n",
       "304309                                Sparse[int64, 0]\n",
       "16076                                 Sparse[int64, 0]\n",
       "373909                                Sparse[int64, 0]\n",
       "35786                                 Sparse[int64, 0]\n",
       "                                            ...       \n",
       "meta_sex                                        object\n",
       "meta_geographic location_latitude              float64\n",
       "meta_geographic location_longitude             float64\n",
       "meta_geographic location_country                object\n",
       "meta_Description                                object\n",
       "Length: 79613, dtype: object"
      ]
     },
     "execution_count": 104,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.dtypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-105-4263561188a0>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mdf\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mto_csv\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"../Human_Subramanian/INPUT_FILES/subramanian_et_al_data.csv\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mindex\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mFalse\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m~/EMAN2/envs/microbiome_toolbox/lib/python3.6/site-packages/pandas/core/generic.py\u001b[0m in \u001b[0;36mto_csv\u001b[0;34m(self, path_or_buf, sep, na_rep, float_format, columns, header, index, index_label, mode, encoding, compression, quoting, quotechar, line_terminator, chunksize, date_format, doublequote, escapechar, decimal, errors)\u001b[0m\n\u001b[1;32m   3168\u001b[0m             \u001b[0mdecimal\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mdecimal\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3169\u001b[0m         )\n\u001b[0;32m-> 3170\u001b[0;31m         \u001b[0mformatter\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msave\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   3171\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3172\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mpath_or_buf\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/EMAN2/envs/microbiome_toolbox/lib/python3.6/site-packages/pandas/io/formats/csvs.py\u001b[0m in \u001b[0;36msave\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    204\u001b[0m             )\n\u001b[1;32m    205\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 206\u001b[0;31m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_save\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    207\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    208\u001b[0m         \u001b[0;32mfinally\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/EMAN2/envs/microbiome_toolbox/lib/python3.6/site-packages/pandas/io/formats/csvs.py\u001b[0m in \u001b[0;36m_save\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    326\u001b[0m                 \u001b[0;32mbreak\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    327\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 328\u001b[0;31m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_save_chunk\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mstart_i\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mend_i\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    329\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    330\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m_save_chunk\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mstart_i\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mint\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mend_i\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mint\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m->\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/EMAN2/envs/microbiome_toolbox/lib/python3.6/site-packages/pandas/io/formats/csvs.py\u001b[0m in \u001b[0;36m_save_chunk\u001b[0;34m(self, start_i, end_i)\u001b[0m\n\u001b[1;32m    344\u001b[0m                 \u001b[0mdecimal\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdecimal\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    345\u001b[0m                 \u001b[0mdate_format\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdate_format\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 346\u001b[0;31m                 \u001b[0mquoting\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mquoting\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    347\u001b[0m             )\n\u001b[1;32m    348\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/EMAN2/envs/microbiome_toolbox/lib/python3.6/site-packages/pandas/core/internals/blocks.py\u001b[0m in \u001b[0;36mto_native_types\u001b[0;34m(self, na_rep, quoting, **kwargs)\u001b[0m\n\u001b[1;32m   1702\u001b[0m         \u001b[0mmask\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0misna\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mvalues\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1703\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1704\u001b[0;31m         \u001b[0mvalues\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0masarray\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mvalues\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mastype\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mobject\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1705\u001b[0m         \u001b[0mvalues\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mmask\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mna_rep\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1706\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/EMAN2/envs/microbiome_toolbox/lib/python3.6/site-packages/pandas/core/arrays/sparse/array.py\u001b[0m in \u001b[0;36mastype\u001b[0;34m(self, dtype, copy)\u001b[0m\n\u001b[1;32m   1068\u001b[0m         \u001b[0;31m# TODO copy=False is broken for astype_nansafe with int -> float, so cannot\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1069\u001b[0m         \u001b[0;31m# passthrough copy keyword: https://github.com/pandas-dev/pandas/issues/34456\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1070\u001b[0;31m         \u001b[0msp_values\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mastype_nansafe\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msp_values\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msubtype\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcopy\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1071\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0msp_values\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msp_values\u001b[0m \u001b[0;32mand\u001b[0m \u001b[0mcopy\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1072\u001b[0m             \u001b[0msp_values\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0msp_values\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcopy\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/EMAN2/envs/microbiome_toolbox/lib/python3.6/site-packages/pandas/core/dtypes/cast.py\u001b[0m in \u001b[0;36mastype_nansafe\u001b[0;34m(arr, dtype, copy, skipna)\u001b[0m\n\u001b[1;32m    993\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0mcopy\u001b[0m \u001b[0;32mor\u001b[0m \u001b[0mis_object_dtype\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0marr\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mor\u001b[0m \u001b[0mis_object_dtype\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdtype\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    994\u001b[0m         \u001b[0;31m# Explicit copy, or required since NumPy can't view from / to object.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 995\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0marr\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mastype\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdtype\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcopy\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    996\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    997\u001b[0m     \u001b[0;32mreturn\u001b[0m \u001b[0marr\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mview\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdtype\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "df.to_csv(\"../Human_Subramanian/INPUT_FILES/subramanian_et_al_data.csv\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "RAW_DATA\t\t  subramanian_et_al_l4.csv  subramanian_et_al_l7.csv\r\n",
      "subramanian_et_al_l2.csv  subramanian_et_al_l5.csv\r\n",
      "subramanian_et_al_l3.csv  subramanian_et_al_l6.csv\r\n"
     ]
    }
   ],
   "source": [
    "!ls ../Human_Subramanian/INPUT_FILES/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 200,
   "metadata": {},
   "outputs": [],
   "source": [
    "file_name = \"../Human_Subramanian/INPUT_FILES/subramanian_et_al_l4.csv\"\n",
    "df_metadata = pd.read_csv(file_name, sep=None, engine='python')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 201,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Unnamed: 0',\n",
       " 'x.sampleid',\n",
       " 'barcodesequence',\n",
       " 'linkerprimersequence',\n",
       " 'personid',\n",
       " 'familyid',\n",
       " 'ena.sampleid',\n",
       " 'ena.libraryname',\n",
       " 'health_analysis_groups',\n",
       " 'tax_id',\n",
       " 'scientific_name',\n",
       " 'common_name',\n",
       " 'anonymized_name',\n",
       " 'sample_title',\n",
       " 'sample_description',\n",
       " 'investigation.type',\n",
       " 'project.name',\n",
       " 'target.gene',\n",
       " 'sequencing.method',\n",
       " 'collection.date',\n",
       " 'host.subject.id',\n",
       " 'age_in_months',\n",
       " 'sex',\n",
       " 'environmental.package',\n",
       " 'geographic.location_latitude',\n",
       " 'geographic.location_longitude',\n",
       " 'geographic.location_country',\n",
       " 'environment_biome',\n",
       " 'environment_feature',\n",
       " 'environment_material',\n",
       " 'description',\n",
       " 'unassigned.other',\n",
       " 'k__archaea.p__crenarchaeota',\n",
       " 'k__archaea.p__euryarchaeota',\n",
       " 'k__bacteria.other',\n",
       " 'k__bacteria.p__ad3',\n",
       " 'k__bacteria.p__actinobacteria',\n",
       " 'k__bacteria.p__bacteroidetes',\n",
       " 'k__bacteria.p__chlamydiae',\n",
       " 'k__bacteria.p__cyanobacteria',\n",
       " 'k__bacteria.p__elusimicrobia',\n",
       " 'k__bacteria.p__firmicutes',\n",
       " 'k__bacteria.p__fusobacteria',\n",
       " 'k__bacteria.p__lentisphaerae',\n",
       " 'k__bacteria.p__planctomycetes',\n",
       " 'k__bacteria.p__proteobacteria',\n",
       " 'k__bacteria.p__spirochaetes',\n",
       " 'k__bacteria.p__synergistetes',\n",
       " 'k__bacteria.p__tm7',\n",
       " 'k__bacteria.p__tenericutes',\n",
       " 'k__bacteria.p__verrucomicrobia',\n",
       " 'k__bacteria.p__wps.2',\n",
       " 'k__bacteria.p__.thermi.',\n",
       " 'unassigned.other.other',\n",
       " 'k__archaea.p__crenarchaeota.c__thaumarchaeota',\n",
       " 'k__archaea.p__euryarchaeota.c__methanobacteria',\n",
       " 'k__bacteria.other.other',\n",
       " 'k__bacteria.p__ad3.c__jg37.ag.4',\n",
       " 'k__bacteria.p__actinobacteria.c__actinobacteria',\n",
       " 'k__bacteria.p__actinobacteria.c__coriobacteriia',\n",
       " 'k__bacteria.p__actinobacteria.c__rubrobacteria',\n",
       " 'k__bacteria.p__actinobacteria.c__thermoleophilia',\n",
       " 'k__bacteria.p__bacteroidetes.c__bacteroidia',\n",
       " 'k__bacteria.p__bacteroidetes.c__flavobacteriia',\n",
       " 'k__bacteria.p__bacteroidetes.c__sphingobacteriia',\n",
       " 'k__bacteria.p__bacteroidetes.c__.saprospirae.',\n",
       " 'k__bacteria.p__chlamydiae.c__chlamydiia',\n",
       " 'k__bacteria.p__cyanobacteria.c__4c0d.2',\n",
       " 'k__bacteria.p__cyanobacteria.c__chloroplast',\n",
       " 'k__bacteria.p__elusimicrobia.c__elusimicrobia',\n",
       " 'k__bacteria.p__firmicutes.other',\n",
       " 'k__bacteria.p__firmicutes.c__bacilli',\n",
       " 'k__bacteria.p__firmicutes.c__clostridia',\n",
       " 'k__bacteria.p__firmicutes.c__erysipelotrichi',\n",
       " 'k__bacteria.p__fusobacteria.c__fusobacteriia',\n",
       " 'k__bacteria.p__lentisphaerae.c__.lentisphaeria.',\n",
       " 'k__bacteria.p__planctomycetes.c__planctomycetia',\n",
       " 'k__bacteria.p__proteobacteria.other',\n",
       " 'k__bacteria.p__proteobacteria.c__alphaproteobacteria',\n",
       " 'k__bacteria.p__proteobacteria.c__betaproteobacteria',\n",
       " 'k__bacteria.p__proteobacteria.c__deltaproteobacteria',\n",
       " 'k__bacteria.p__proteobacteria.c__epsilonproteobacteria',\n",
       " 'k__bacteria.p__proteobacteria.c__gammaproteobacteria',\n",
       " 'k__bacteria.p__spirochaetes.c__spirochaetes',\n",
       " 'k__bacteria.p__spirochaetes.c__.brachyspirae.',\n",
       " 'k__bacteria.p__synergistetes.c__synergistia',\n",
       " 'k__bacteria.p__tm7.c__tm7.3',\n",
       " 'k__bacteria.p__tenericutes.c__ck.1c4.19',\n",
       " 'k__bacteria.p__tenericutes.c__mollicutes',\n",
       " 'k__bacteria.p__tenericutes.c__rf3',\n",
       " 'k__bacteria.p__verrucomicrobia.c__opitutae',\n",
       " 'k__bacteria.p__verrucomicrobia.c__verruco.5',\n",
       " 'k__bacteria.p__verrucomicrobia.c__verrucomicrobiae',\n",
       " 'k__bacteria.p__wps.2.c__',\n",
       " 'k__bacteria.p__.thermi..c__deinococci',\n",
       " 'unassigned.other.other.other',\n",
       " 'k__archaea.p__crenarchaeota.c__thaumarchaeota.o__nitrososphaerales',\n",
       " 'k__archaea.p__euryarchaeota.c__methanobacteria.o__methanobacteriales',\n",
       " 'k__bacteria.other.other.other',\n",
       " 'k__bacteria.p__ad3.c__jg37.ag.4.o__',\n",
       " 'k__bacteria.p__actinobacteria.c__actinobacteria.o__actinomycetales',\n",
       " 'k__bacteria.p__actinobacteria.c__actinobacteria.o__bifidobacteriales',\n",
       " 'k__bacteria.p__actinobacteria.c__coriobacteriia.o__coriobacteriales',\n",
       " 'k__bacteria.p__actinobacteria.c__rubrobacteria.o__rubrobacterales',\n",
       " 'k__bacteria.p__actinobacteria.c__thermoleophilia.o__gaiellales',\n",
       " 'k__bacteria.p__actinobacteria.c__thermoleophilia.o__solirubrobacterales',\n",
       " 'k__bacteria.p__bacteroidetes.c__bacteroidia.o__bacteroidales',\n",
       " 'k__bacteria.p__bacteroidetes.c__flavobacteriia.o__flavobacteriales',\n",
       " 'k__bacteria.p__bacteroidetes.c__sphingobacteriia.o__sphingobacteriales',\n",
       " 'k__bacteria.p__bacteroidetes.c__.saprospirae..o__.saprospirales.',\n",
       " 'k__bacteria.p__chlamydiae.c__chlamydiia.o__chlamydiales',\n",
       " 'k__bacteria.p__cyanobacteria.c__4c0d.2.o__ys2',\n",
       " 'k__bacteria.p__cyanobacteria.c__chloroplast.o__streptophyta',\n",
       " 'k__bacteria.p__elusimicrobia.c__elusimicrobia.o__elusimicrobiales',\n",
       " 'k__bacteria.p__firmicutes.other.other',\n",
       " 'k__bacteria.p__firmicutes.c__bacilli.other',\n",
       " 'k__bacteria.p__firmicutes.c__bacilli.o__bacillales',\n",
       " 'k__bacteria.p__firmicutes.c__bacilli.o__gemellales',\n",
       " 'k__bacteria.p__firmicutes.c__bacilli.o__lactobacillales',\n",
       " 'k__bacteria.p__firmicutes.c__bacilli.o__turicibacterales',\n",
       " 'k__bacteria.p__firmicutes.c__clostridia.o__clostridiales',\n",
       " 'k__bacteria.p__firmicutes.c__erysipelotrichi.o__erysipelotrichales',\n",
       " 'k__bacteria.p__fusobacteria.c__fusobacteriia.o__fusobacteriales',\n",
       " 'k__bacteria.p__lentisphaerae.c__.lentisphaeria..o__victivallales',\n",
       " 'k__bacteria.p__planctomycetes.c__planctomycetia.o__pirellulales',\n",
       " 'k__bacteria.p__planctomycetes.c__planctomycetia.o__planctomycetales',\n",
       " 'k__bacteria.p__proteobacteria.other.other',\n",
       " 'k__bacteria.p__proteobacteria.c__alphaproteobacteria.o__',\n",
       " 'k__bacteria.p__proteobacteria.c__alphaproteobacteria.o__caulobacterales',\n",
       " 'k__bacteria.p__proteobacteria.c__alphaproteobacteria.o__rf32',\n",
       " 'k__bacteria.p__proteobacteria.c__alphaproteobacteria.o__rhizobiales',\n",
       " 'k__bacteria.p__proteobacteria.c__alphaproteobacteria.o__rhodobacterales',\n",
       " 'k__bacteria.p__proteobacteria.c__alphaproteobacteria.o__rhodospirillales',\n",
       " 'k__bacteria.p__proteobacteria.c__alphaproteobacteria.o__rickettsiales',\n",
       " 'k__bacteria.p__proteobacteria.c__alphaproteobacteria.o__sphingomonadales',\n",
       " 'k__bacteria.p__proteobacteria.c__betaproteobacteria.other',\n",
       " 'k__bacteria.p__proteobacteria.c__betaproteobacteria.o__burkholderiales',\n",
       " 'k__bacteria.p__proteobacteria.c__betaproteobacteria.o__methylophilales',\n",
       " 'k__bacteria.p__proteobacteria.c__betaproteobacteria.o__neisseriales',\n",
       " 'k__bacteria.p__proteobacteria.c__betaproteobacteria.o__rhodocyclales',\n",
       " 'k__bacteria.p__proteobacteria.c__deltaproteobacteria.o__bdellovibrionales',\n",
       " 'k__bacteria.p__proteobacteria.c__deltaproteobacteria.o__desulfobacterales',\n",
       " 'k__bacteria.p__proteobacteria.c__deltaproteobacteria.o__desulfovibrionales',\n",
       " 'k__bacteria.p__proteobacteria.c__deltaproteobacteria.o__myxococcales',\n",
       " 'k__bacteria.p__proteobacteria.c__epsilonproteobacteria.o__campylobacterales',\n",
       " 'k__bacteria.p__proteobacteria.c__gammaproteobacteria.other',\n",
       " 'k__bacteria.p__proteobacteria.c__gammaproteobacteria.o__aeromonadales',\n",
       " 'k__bacteria.p__proteobacteria.c__gammaproteobacteria.o__alteromonadales',\n",
       " 'k__bacteria.p__proteobacteria.c__gammaproteobacteria.o__cardiobacteriales',\n",
       " 'k__bacteria.p__proteobacteria.c__gammaproteobacteria.o__enterobacteriales',\n",
       " 'k__bacteria.p__proteobacteria.c__gammaproteobacteria.o__legionellales',\n",
       " 'k__bacteria.p__proteobacteria.c__gammaproteobacteria.o__oceanospirillales',\n",
       " 'k__bacteria.p__proteobacteria.c__gammaproteobacteria.o__pasteurellales',\n",
       " 'k__bacteria.p__proteobacteria.c__gammaproteobacteria.o__pseudomonadales',\n",
       " 'k__bacteria.p__proteobacteria.c__gammaproteobacteria.o__xanthomonadales',\n",
       " 'k__bacteria.p__spirochaetes.c__spirochaetes.o__spirochaetales',\n",
       " 'k__bacteria.p__spirochaetes.c__.brachyspirae..o__.brachyspirales.',\n",
       " 'k__bacteria.p__synergistetes.c__synergistia.o__synergistales',\n",
       " 'k__bacteria.p__tm7.c__tm7.3.o__',\n",
       " 'k__bacteria.p__tm7.c__tm7.3.o__cw040',\n",
       " 'k__bacteria.p__tm7.c__tm7.3.o__i025',\n",
       " 'k__bacteria.p__tenericutes.c__ck.1c4.19.o__',\n",
       " 'k__bacteria.p__tenericutes.c__mollicutes.o__',\n",
       " 'k__bacteria.p__tenericutes.c__mollicutes.o__anaeroplasmatales',\n",
       " 'k__bacteria.p__tenericutes.c__mollicutes.o__mycoplasmatales',\n",
       " 'k__bacteria.p__tenericutes.c__mollicutes.o__rf39',\n",
       " 'k__bacteria.p__tenericutes.c__rf3.o__ml615j.28',\n",
       " 'k__bacteria.p__verrucomicrobia.c__opitutae.o__.cerasicoccales.',\n",
       " 'k__bacteria.p__verrucomicrobia.c__verruco.5.o__wchb1.41',\n",
       " 'k__bacteria.p__verrucomicrobia.c__verrucomicrobiae.o__verrucomicrobiales',\n",
       " 'k__bacteria.p__wps.2.c__.o__',\n",
       " 'k__bacteria.p__.thermi..c__deinococci.o__deinococcales',\n",
       " 'study']"
      ]
     },
     "execution_count": 201,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_metadata.columns.tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 202,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['BANG_HLTHY', 'HS1_Pool6', 'HS1_Pool5', 'HS1_Pool7', 'HS1_Pool4',\n",
       "       'ACMAL_3_SS2  and  ACMAL_3_SS4', 'SS3', 'SS1', 'Twins34',\n",
       "       'Twins12', 'Twins56_run1 and Twins56_run2', 'Twins78',\n",
       "       'TwinsCP2_run1 and TwinsCP2_run2', 'TwinsRX',\n",
       "       'ACMAL_1_run1 and ACMAL_1_run2', 'ACMAL_2_run1 and ACMAL_2_run2',\n",
       "       'GGS89', 'GGS7', 'GGS6', 'GGS3'], dtype=object)"
      ]
     },
     "execution_count": 202,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_metadata[\"ena.libraryname\"].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 203,
   "metadata": {},
   "outputs": [],
   "source": [
    "# TODO: handle group and reference_group"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 204,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_metadata.age_in_months *= 30"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 205,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# id_columns = [\"FamilyID\", \"ENA.SampleID\", \"tax_id\"]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 206,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_metadata = df_metadata.rename(\n",
    "    {\"x.sampleid\": \"sampleID\", \n",
    "   \"personid\": \"subjectID\",\n",
    "   \"age_in_months\": \"age_at_collection\",\n",
    "#     \"reference_group\": \n",
    "     \"health_analysis_groups\": \"group\"\n",
    "    }, axis=1)   \n",
    "\n",
    "       \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 207,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "linkerprimersequence has only 1 unique value [nan]\n",
      "tax_id has only 1 unique value [408170]\n",
      "scientific_name has only 1 unique value ['human gut metagenome']\n",
      "common_name has only 1 unique value ['human gut metagenome']\n",
      "sample_description has only 1 unique value ['human gut metagenome']\n",
      "investigation.type has only 1 unique value ['mimarks-survey']\n",
      "project.name has only 1 unique value ['Subramanian maturity']\n",
      "target.gene has only 1 unique value ['V4']\n",
      "sequencing.method has only 1 unique value ['Illumina MiSeq']\n",
      "collection.date has only 1 unique value ['None']\n",
      "environmental.package has only 1 unique value ['human-gut']\n",
      "environment_biome has only 1 unique value ['ENVO:human-associated habitat']\n",
      "environment_feature has only 1 unique value ['ENVO:human-associated habitat']\n",
      "environment_material has only 1 unique value ['ENVO:feces']\n",
      "study has only 1 unique value ['RM_Bangladesh']\n"
     ]
    }
   ],
   "source": [
    "columns_to_remove = [\"Unnamed: 0\",\n",
    "                     \"ena.sampleid\", \n",
    "                     \"anonymized_name\",\n",
    "                     \"sample_title\",\n",
    "                     \"host subject id\",\n",
    "                     \"description\"]\n",
    "for col in df_metadata.columns:\n",
    "    if len(df_metadata[col].unique()) == 1:\n",
    "        print(col, \"has only 1 unique value\", df_metadata[col].unique())\n",
    "        df_metadata.drop(col, inplace=True, axis=1)\n",
    "    elif col in columns_to_remove:\n",
    "        df_metadata.drop(col, inplace=True, axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 208,
   "metadata": {},
   "outputs": [],
   "source": [
    "other_names = {}\n",
    "for col1 in set(df_metadata.columns).difference(set(MicrobiomeDataset.specific_columns)):\n",
    "    if \"bacteria\" in col1:\n",
    "        if not col1.startswith(\"bacteria_\"):\n",
    "            other_names[col1] = f\"bacteria_{col1}\"\n",
    "    else:\n",
    "        other_names[col1] = f\"meta_{col1}\"\n",
    "    \n",
    "df_metadata = df_metadata.rename(other_names, axis=1) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 209,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['sampleID', 'meta_barcodesequence', 'subjectID', 'meta_familyid',\n",
       "       'meta_ena.libraryname', 'group', 'meta_host.subject.id',\n",
       "       'age_at_collection', 'meta_sex', 'meta_geographic.location_latitude',\n",
       "       ...\n",
       "       'bacteria_k__bacteria.p__tenericutes.c__mollicutes.o__',\n",
       "       'bacteria_k__bacteria.p__tenericutes.c__mollicutes.o__anaeroplasmatales',\n",
       "       'bacteria_k__bacteria.p__tenericutes.c__mollicutes.o__mycoplasmatales',\n",
       "       'bacteria_k__bacteria.p__tenericutes.c__mollicutes.o__rf39',\n",
       "       'bacteria_k__bacteria.p__tenericutes.c__rf3.o__ml615j.28',\n",
       "       'bacteria_k__bacteria.p__verrucomicrobia.c__opitutae.o__.cerasicoccales.',\n",
       "       'bacteria_k__bacteria.p__verrucomicrobia.c__verruco.5.o__wchb1.41',\n",
       "       'bacteria_k__bacteria.p__verrucomicrobia.c__verrucomicrobiae.o__verrucomicrobiales',\n",
       "       'bacteria_k__bacteria.p__wps.2.c__.o__',\n",
       "       'bacteria_k__bacteria.p__.thermi..c__deinococci.o__deinococcales'],\n",
       "      dtype='object', length=153)"
      ]
     },
     "execution_count": 209,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_metadata.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 218,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['sampleID', 'meta_barcodesequence', 'subjectID', 'meta_familyid',\n",
       "       'meta_ena.libraryname', 'group', 'meta_host.subject.id',\n",
       "       'age_at_collection', 'meta_sex', 'meta_geographic.location_latitude',\n",
       "       ...\n",
       "       'bacteria_k__bacteria.p__tenericutes.c__mollicutes.o__',\n",
       "       'bacteria_k__bacteria.p__tenericutes.c__mollicutes.o__anaeroplasmatales',\n",
       "       'bacteria_k__bacteria.p__tenericutes.c__mollicutes.o__mycoplasmatales',\n",
       "       'bacteria_k__bacteria.p__tenericutes.c__mollicutes.o__rf39',\n",
       "       'bacteria_k__bacteria.p__tenericutes.c__rf3.o__ml615j.28',\n",
       "       'bacteria_k__bacteria.p__verrucomicrobia.c__opitutae.o__.cerasicoccales.',\n",
       "       'bacteria_k__bacteria.p__verrucomicrobia.c__verruco.5.o__wchb1.41',\n",
       "       'bacteria_k__bacteria.p__verrucomicrobia.c__verrucomicrobiae.o__verrucomicrobiales',\n",
       "       'bacteria_k__bacteria.p__wps.2.c__.o__',\n",
       "       'bacteria_k__bacteria.p__.thermi..c__deinococci.o__deinococcales'],\n",
       "      dtype='object', length=153)"
      ]
     },
     "execution_count": 218,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_metadata.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 211,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sampleID</th>\n",
       "      <th>meta_barcodesequence</th>\n",
       "      <th>subjectID</th>\n",
       "      <th>meta_familyid</th>\n",
       "      <th>meta_ena.libraryname</th>\n",
       "      <th>group</th>\n",
       "      <th>meta_host.subject.id</th>\n",
       "      <th>age_at_collection</th>\n",
       "      <th>meta_sex</th>\n",
       "      <th>meta_geographic.location_latitude</th>\n",
       "      <th>...</th>\n",
       "      <th>bacteria_k__bacteria.p__tenericutes.c__mollicutes.o__</th>\n",
       "      <th>bacteria_k__bacteria.p__tenericutes.c__mollicutes.o__anaeroplasmatales</th>\n",
       "      <th>bacteria_k__bacteria.p__tenericutes.c__mollicutes.o__mycoplasmatales</th>\n",
       "      <th>bacteria_k__bacteria.p__tenericutes.c__mollicutes.o__rf39</th>\n",
       "      <th>bacteria_k__bacteria.p__tenericutes.c__rf3.o__ml615j.28</th>\n",
       "      <th>bacteria_k__bacteria.p__verrucomicrobia.c__opitutae.o__.cerasicoccales.</th>\n",
       "      <th>bacteria_k__bacteria.p__verrucomicrobia.c__verruco.5.o__wchb1.41</th>\n",
       "      <th>bacteria_k__bacteria.p__verrucomicrobia.c__verrucomicrobiae.o__verrucomicrobiales</th>\n",
       "      <th>bacteria_k__bacteria.p__wps.2.c__.o__</th>\n",
       "      <th>bacteria_k__bacteria.p__.thermi..c__deinococci.o__deinococcales</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Bgsng7035.m1</td>\n",
       "      <td>GGATAGCCACTTC</td>\n",
       "      <td>Bgsng7035</td>\n",
       "      <td>Bgsng7035</td>\n",
       "      <td>BANG_HLTHY</td>\n",
       "      <td>Healthy Singletons</td>\n",
       "      <td>Bgsng7035</td>\n",
       "      <td>6.0</td>\n",
       "      <td>male</td>\n",
       "      <td>23.933</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Bgsng7035.m2</td>\n",
       "      <td>TTGGGCGTGAACT</td>\n",
       "      <td>Bgsng7035</td>\n",
       "      <td>Bgsng7035</td>\n",
       "      <td>BANG_HLTHY</td>\n",
       "      <td>Healthy Singletons</td>\n",
       "      <td>Bgsng7035</td>\n",
       "      <td>36.0</td>\n",
       "      <td>male</td>\n",
       "      <td>23.933</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Bgsng7035.m3</td>\n",
       "      <td>TAACAAGGAACGC</td>\n",
       "      <td>Bgsng7035</td>\n",
       "      <td>Bgsng7035</td>\n",
       "      <td>BANG_HLTHY</td>\n",
       "      <td>Healthy Singletons</td>\n",
       "      <td>Bgsng7035</td>\n",
       "      <td>57.0</td>\n",
       "      <td>male</td>\n",
       "      <td>23.933</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Bgsng7035.m4</td>\n",
       "      <td>TCTGCAGTTGGAC</td>\n",
       "      <td>Bgsng7035</td>\n",
       "      <td>Bgsng7035</td>\n",
       "      <td>BANG_HLTHY</td>\n",
       "      <td>Healthy Singletons</td>\n",
       "      <td>Bgsng7035</td>\n",
       "      <td>87.0</td>\n",
       "      <td>male</td>\n",
       "      <td>23.933</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Bgsng7035.m5</td>\n",
       "      <td>CTCGGCAATTATG</td>\n",
       "      <td>Bgsng7035</td>\n",
       "      <td>Bgsng7035</td>\n",
       "      <td>BANG_HLTHY</td>\n",
       "      <td>Healthy Singletons</td>\n",
       "      <td>Bgsng7035</td>\n",
       "      <td>108.0</td>\n",
       "      <td>male</td>\n",
       "      <td>23.933</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 153 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       sampleID meta_barcodesequence  subjectID meta_familyid  \\\n",
       "0  Bgsng7035.m1        GGATAGCCACTTC  Bgsng7035     Bgsng7035   \n",
       "1  Bgsng7035.m2        TTGGGCGTGAACT  Bgsng7035     Bgsng7035   \n",
       "2  Bgsng7035.m3        TAACAAGGAACGC  Bgsng7035     Bgsng7035   \n",
       "3  Bgsng7035.m4        TCTGCAGTTGGAC  Bgsng7035     Bgsng7035   \n",
       "4  Bgsng7035.m5        CTCGGCAATTATG  Bgsng7035     Bgsng7035   \n",
       "\n",
       "  meta_ena.libraryname               group meta_host.subject.id  \\\n",
       "0           BANG_HLTHY  Healthy Singletons            Bgsng7035   \n",
       "1           BANG_HLTHY  Healthy Singletons            Bgsng7035   \n",
       "2           BANG_HLTHY  Healthy Singletons            Bgsng7035   \n",
       "3           BANG_HLTHY  Healthy Singletons            Bgsng7035   \n",
       "4           BANG_HLTHY  Healthy Singletons            Bgsng7035   \n",
       "\n",
       "   age_at_collection meta_sex  meta_geographic.location_latitude  ...  \\\n",
       "0                6.0     male                             23.933  ...   \n",
       "1               36.0     male                             23.933  ...   \n",
       "2               57.0     male                             23.933  ...   \n",
       "3               87.0     male                             23.933  ...   \n",
       "4              108.0     male                             23.933  ...   \n",
       "\n",
       "   bacteria_k__bacteria.p__tenericutes.c__mollicutes.o__  \\\n",
       "0                                                0.0       \n",
       "1                                                0.0       \n",
       "2                                                0.0       \n",
       "3                                                0.0       \n",
       "4                                                0.0       \n",
       "\n",
       "  bacteria_k__bacteria.p__tenericutes.c__mollicutes.o__anaeroplasmatales  \\\n",
       "0                                                0.0                       \n",
       "1                                                0.0                       \n",
       "2                                                0.0                       \n",
       "3                                                0.0                       \n",
       "4                                                0.0                       \n",
       "\n",
       "   bacteria_k__bacteria.p__tenericutes.c__mollicutes.o__mycoplasmatales  \\\n",
       "0                                                0.0                      \n",
       "1                                                0.0                      \n",
       "2                                                0.0                      \n",
       "3                                                0.0                      \n",
       "4                                                0.0                      \n",
       "\n",
       "   bacteria_k__bacteria.p__tenericutes.c__mollicutes.o__rf39  \\\n",
       "0                                                0.0           \n",
       "1                                                0.0           \n",
       "2                                                0.0           \n",
       "3                                                0.0           \n",
       "4                                                0.0           \n",
       "\n",
       "   bacteria_k__bacteria.p__tenericutes.c__rf3.o__ml615j.28  \\\n",
       "0                                                0.0         \n",
       "1                                                0.0         \n",
       "2                                                0.0         \n",
       "3                                                0.0         \n",
       "4                                                0.0         \n",
       "\n",
       "   bacteria_k__bacteria.p__verrucomicrobia.c__opitutae.o__.cerasicoccales.  \\\n",
       "0                                                0.0                         \n",
       "1                                                0.0                         \n",
       "2                                                0.0                         \n",
       "3                                                0.0                         \n",
       "4                                                0.0                         \n",
       "\n",
       "   bacteria_k__bacteria.p__verrucomicrobia.c__verruco.5.o__wchb1.41  \\\n",
       "0                                                0.0                  \n",
       "1                                                0.0                  \n",
       "2                                                0.0                  \n",
       "3                                                0.0                  \n",
       "4                                                0.0                  \n",
       "\n",
       "   bacteria_k__bacteria.p__verrucomicrobia.c__verrucomicrobiae.o__verrucomicrobiales  \\\n",
       "0                                                0.0                                   \n",
       "1                                                0.0                                   \n",
       "2                                                0.0                                   \n",
       "3                                                0.0                                   \n",
       "4                                                0.0                                   \n",
       "\n",
       "   bacteria_k__bacteria.p__wps.2.c__.o__  \\\n",
       "0                                    0.0   \n",
       "1                                    0.0   \n",
       "2                                    0.0   \n",
       "3                                    0.0   \n",
       "4                                    0.0   \n",
       "\n",
       "   bacteria_k__bacteria.p__.thermi..c__deinococci.o__deinococcales  \n",
       "0                                                0.0                \n",
       "1                                                0.0                \n",
       "2                                                0.0                \n",
       "3                                                0.0                \n",
       "4                                                0.0                \n",
       "\n",
       "[5 rows x 153 columns]"
      ]
     },
     "execution_count": 211,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_metadata.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 212,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_metadata.to_csv(\"../Human_Subramanian/INPUT_FILES/subramanian_et_al_l2_website.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
