{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Reference:\n",
    "- [Approximating Observed Microbiome Data with microbiomeDASim in R](https://github.com/williazo/microbiomeDASim/blob/master/inst/script/mouse_microbiome_approximation.ipynb)\n",
    "\n",
    "Note: the version of R 4.0.3 at least!\n",
    "\n",
    "> This dataset is stored as an MRexperiment object with assay data collected at the OTU level. The raw counts represent over 10,000 OTUs that were sequences from a total of 139 samples. These samples represent repeated measurements taken on 12 gnotobiotic mice. All mice were fed the same low-fat, plant polysaccharide–rich diet for the first 21 days of the study. At this point 6 of the mice were then switched to a high-fat, high-sugar “Western” diet. The subseqeuent changes in the microbial community were then observed over a follow-up of roughly 60 days.\n",
    "> Many of the OTUs are low frequency obserations that do not match out multivariate normal distributional assumptions, especially given our limited sample size (n=12). Therefore, we restrict our analysis to genus level counts rather than OTU level counts, which reduces the feature size from 10,172 to 61.\n",
    "> We further apply presence and depth filters, and then log normalize the counts to represent our simulated outcome of interest.\n",
    "\n",
    "So, the taxa values are log-normalized counts."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "```R\n",
    "# install package\n",
    "if (!requireNamespace(\"BiocManager\", quietly = TRUE))\n",
    "  install.packages(\"BiocManager\")\n",
    "BiocManager::install(\"metagenomeSeq\")\n",
    "\n",
    "# import mouseData from metagenomeSeq\n",
    "library(metagenomeSeq)\n",
    "data(\"mouseData\")\n",
    "\n",
    "# aggregating the counts to the genus level\n",
    "genus_mouseData <- metagenomeSeq::aggTax(mouseData, lvl=\"genus\")\n",
    "\n",
    "# additional count filters\n",
    "genus_mouseData <- filterData(genus_mouseData, present = 10, depth = 1000)\n",
    "g_lnorm_mat <- MRcounts(genus_mouseData, norm=TRUE, log=TRUE)\n",
    "\n",
    "# metadata\n",
    "pd_mouseData <- pData(genus_mouseData)\n",
    "\n",
    "# save metadata file\n",
    "write.csv(pd_mouseData, 'C://Users//RDBanjacJe//OneDrive - NESTLE//Jelena//ELMToolBox-Project//notebooks//mousedata_metadata.csv')\n",
    "# save features data\n",
    "write.csv(g_lnorm_mat, 'C://Users//RDBanjacJe//OneDrive - NESTLE//Jelena//ELMToolBox-Project//notebooks//mousedata_features.csv')\n",
    "\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "sys.path.append(\"C://Users//RDBanjacJe//Desktop//ELMToolBox\") \n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import seaborn as sns\n",
    "from elmtoolbox.preprocessing import dataset_bacteria_abundances"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(137, 5)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mouseID</th>\n",
       "      <th>date</th>\n",
       "      <th>diet</th>\n",
       "      <th>relativeTime</th>\n",
       "      <th>status</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>PM1:20080108</th>\n",
       "      <td>PM1</td>\n",
       "      <td>2008-01-08</td>\n",
       "      <td>BK</td>\n",
       "      <td>22</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PM1:20080114</th>\n",
       "      <td>PM1</td>\n",
       "      <td>2008-01-14</td>\n",
       "      <td>BK</td>\n",
       "      <td>28</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PM1:20071211</th>\n",
       "      <td>PM1</td>\n",
       "      <td>2007-12-11</td>\n",
       "      <td>BK</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PM1:20080121</th>\n",
       "      <td>PM1</td>\n",
       "      <td>2008-01-21</td>\n",
       "      <td>BK</td>\n",
       "      <td>35</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PM1:20071217</th>\n",
       "      <td>PM1</td>\n",
       "      <td>2007-12-17</td>\n",
       "      <td>BK</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             mouseID        date diet  relativeTime  status\n",
       "PM1:20080108     PM1  2008-01-08   BK            22       0\n",
       "PM1:20080114     PM1  2008-01-14   BK            28       0\n",
       "PM1:20071211     PM1  2007-12-11   BK             0       0\n",
       "PM1:20080121     PM1  2008-01-21   BK            35       0\n",
       "PM1:20071217     PM1  2007-12-17   BK             6       0"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# We then combine the observed outcome with phenotype data on the samples which includes their ID, time since study began\n",
    "df_metadata = pd.read_csv(\"../INPUT_FILES/mousedata_metadata.csv\", index_col=0)\n",
    "print(df_metadata.shape)\n",
    "df_metadata.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(137, 35)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>g__Akkermansia</th>\n",
       "      <th>g__Alistipes</th>\n",
       "      <th>g__Anaerofilum</th>\n",
       "      <th>g__Anaerofustis</th>\n",
       "      <th>g__Anaerostipes</th>\n",
       "      <th>g__Anaerotruncus</th>\n",
       "      <th>g__Anaerovorax</th>\n",
       "      <th>g__Bacteroides</th>\n",
       "      <th>g__Bilophila</th>\n",
       "      <th>g__Bryantella</th>\n",
       "      <th>...</th>\n",
       "      <th>g__Mogibacterium</th>\n",
       "      <th>g__nan</th>\n",
       "      <th>g__Parabacteroides</th>\n",
       "      <th>g__Prevotella</th>\n",
       "      <th>g__Roseburia</th>\n",
       "      <th>g__RuminococcaceaeIncertaeSedis</th>\n",
       "      <th>g__Ruminococcus</th>\n",
       "      <th>g__Subdoligranulum</th>\n",
       "      <th>g__Sutterella</th>\n",
       "      <th>g__Turicibacter</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>PM1:20080108</th>\n",
       "      <td>5.198270</td>\n",
       "      <td>8.861973</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>6.178487</td>\n",
       "      <td>6.178487</td>\n",
       "      <td>5.198270</td>\n",
       "      <td>13.537929</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>7.971544</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>14.360622</td>\n",
       "      <td>10.713877</td>\n",
       "      <td>14.109796</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>7.488414</td>\n",
       "      <td>7.750109</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PM1:20080114</th>\n",
       "      <td>8.163470</td>\n",
       "      <td>8.685500</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>5.198270</td>\n",
       "      <td>6.496426</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>14.011826</td>\n",
       "      <td>6.178487</td>\n",
       "      <td>8.916306</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>14.508325</td>\n",
       "      <td>11.041757</td>\n",
       "      <td>8.163470</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>6.756795</td>\n",
       "      <td>9.408481</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>4.237039</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PM1:20071211</th>\n",
       "      <td>8.536715</td>\n",
       "      <td>6.230243</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>6.230243</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>12.998976</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>14.468348</td>\n",
       "      <td>11.044470</td>\n",
       "      <td>13.395906</td>\n",
       "      <td>6.230243</td>\n",
       "      <td>7.220602</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>6.230243</td>\n",
       "      <td>5.249333</td>\n",
       "      <td>5.249333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PM1:20080121</th>\n",
       "      <td>6.658211</td>\n",
       "      <td>8.647458</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>6.658211</td>\n",
       "      <td>5.101538</td>\n",
       "      <td>5.101538</td>\n",
       "      <td>14.394351</td>\n",
       "      <td>6.080373</td>\n",
       "      <td>8.968667</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>14.971828</td>\n",
       "      <td>10.942270</td>\n",
       "      <td>11.451726</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>7.872418</td>\n",
       "      <td>9.382984</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>6.658211</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PM1:20071217</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>9.300649</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.249333</td>\n",
       "      <td>5.249333</td>\n",
       "      <td>5.249333</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>13.803455</td>\n",
       "      <td>8.215758</td>\n",
       "      <td>5.249333</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>14.718745</td>\n",
       "      <td>12.505766</td>\n",
       "      <td>15.998821</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>7.220602</td>\n",
       "      <td>8.023806</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>5.249333</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 35 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              g__Akkermansia  g__Alistipes  g__Anaerofilum  g__Anaerofustis  \\\n",
       "PM1:20080108        5.198270      8.861973             0.0         0.000000   \n",
       "PM1:20080114        8.163470      8.685500             0.0         0.000000   \n",
       "PM1:20071211        8.536715      6.230243             0.0         0.000000   \n",
       "PM1:20080121        6.658211      8.647458             0.0         0.000000   \n",
       "PM1:20071217        0.000000      9.300649             0.0         5.249333   \n",
       "\n",
       "              g__Anaerostipes  g__Anaerotruncus  g__Anaerovorax  \\\n",
       "PM1:20080108         6.178487          6.178487        5.198270   \n",
       "PM1:20080114         5.198270          6.496426        0.000000   \n",
       "PM1:20071211         6.230243          0.000000        0.000000   \n",
       "PM1:20080121         6.658211          5.101538        5.101538   \n",
       "PM1:20071217         5.249333          5.249333        0.000000   \n",
       "\n",
       "              g__Bacteroides  g__Bilophila  g__Bryantella  ...  \\\n",
       "PM1:20080108       13.537929      0.000000       7.971544  ...   \n",
       "PM1:20080114       14.011826      6.178487       8.916306  ...   \n",
       "PM1:20071211       12.998976      0.000000       0.000000  ...   \n",
       "PM1:20080121       14.394351      6.080373       8.968667  ...   \n",
       "PM1:20071217       13.803455      8.215758       5.249333  ...   \n",
       "\n",
       "              g__Mogibacterium     g__nan  g__Parabacteroides  g__Prevotella  \\\n",
       "PM1:20080108               0.0  14.360622           10.713877      14.109796   \n",
       "PM1:20080114               0.0  14.508325           11.041757       8.163470   \n",
       "PM1:20071211               0.0  14.468348           11.044470      13.395906   \n",
       "PM1:20080121               0.0  14.971828           10.942270      11.451726   \n",
       "PM1:20071217               0.0  14.718745           12.505766      15.998821   \n",
       "\n",
       "              g__Roseburia  g__RuminococcaceaeIncertaeSedis  g__Ruminococcus  \\\n",
       "PM1:20080108      0.000000                         7.488414         7.750109   \n",
       "PM1:20080114      0.000000                         6.756795         9.408481   \n",
       "PM1:20071211      6.230243                         7.220602         0.000000   \n",
       "PM1:20080121      0.000000                         7.872418         9.382984   \n",
       "PM1:20071217      0.000000                         7.220602         8.023806   \n",
       "\n",
       "              g__Subdoligranulum  g__Sutterella  g__Turicibacter  \n",
       "PM1:20080108            0.000000       0.000000         0.000000  \n",
       "PM1:20080114            0.000000       4.237039         0.000000  \n",
       "PM1:20071211            6.230243       5.249333         5.249333  \n",
       "PM1:20080121            0.000000       6.658211         0.000000  \n",
       "PM1:20071217            0.000000       5.249333         0.000000  \n",
       "\n",
       "[5 rows x 35 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Many of the OTUs are low frequency obserations that do not match out multivariate normal distributional assumptions, especially given our limited sample size (n=12). Therefore, we restrict our analysis to genus level counts rather than OTU level counts, which reduces the feature size from 10,172 to 61.\n",
    "# We further apply presence and depth filters, and then log normalize the counts to represent our simulated outcome of interest.\n",
    "df_features = pd.read_csv(\"../INPUT_FILES/mousedata_features.csv\", index_col=0)\n",
    "df_features = df_features.T\n",
    "mapper = {}\n",
    "for c in df_features.columns:\n",
    "    mapper[c] = f\"g__{c}\"\n",
    "df_features = df_features.rename(mapper=mapper, axis=1)\n",
    "\n",
    "print(df_features.shape)\n",
    "df_features.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(137, 41)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sampleID</th>\n",
       "      <th>g__Akkermansia</th>\n",
       "      <th>g__Alistipes</th>\n",
       "      <th>g__Anaerofilum</th>\n",
       "      <th>g__Anaerofustis</th>\n",
       "      <th>g__Anaerostipes</th>\n",
       "      <th>g__Anaerotruncus</th>\n",
       "      <th>g__Anaerovorax</th>\n",
       "      <th>g__Bacteroides</th>\n",
       "      <th>g__Bilophila</th>\n",
       "      <th>...</th>\n",
       "      <th>g__RuminococcaceaeIncertaeSedis</th>\n",
       "      <th>g__Ruminococcus</th>\n",
       "      <th>g__Subdoligranulum</th>\n",
       "      <th>g__Sutterella</th>\n",
       "      <th>g__Turicibacter</th>\n",
       "      <th>mouseID</th>\n",
       "      <th>date</th>\n",
       "      <th>diet</th>\n",
       "      <th>relativeTime</th>\n",
       "      <th>status</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PM1:20080108</td>\n",
       "      <td>5.198270</td>\n",
       "      <td>8.861973</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>6.178487</td>\n",
       "      <td>6.178487</td>\n",
       "      <td>5.198270</td>\n",
       "      <td>13.537929</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>7.488414</td>\n",
       "      <td>7.750109</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>PM1</td>\n",
       "      <td>2008-01-08</td>\n",
       "      <td>BK</td>\n",
       "      <td>22</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PM1:20080114</td>\n",
       "      <td>8.163470</td>\n",
       "      <td>8.685500</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>5.198270</td>\n",
       "      <td>6.496426</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>14.011826</td>\n",
       "      <td>6.178487</td>\n",
       "      <td>...</td>\n",
       "      <td>6.756795</td>\n",
       "      <td>9.408481</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>4.237039</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>PM1</td>\n",
       "      <td>2008-01-14</td>\n",
       "      <td>BK</td>\n",
       "      <td>28</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PM1:20071211</td>\n",
       "      <td>8.536715</td>\n",
       "      <td>6.230243</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>6.230243</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>12.998976</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>7.220602</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>6.230243</td>\n",
       "      <td>5.249333</td>\n",
       "      <td>5.249333</td>\n",
       "      <td>PM1</td>\n",
       "      <td>2007-12-11</td>\n",
       "      <td>BK</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PM1:20080121</td>\n",
       "      <td>6.658211</td>\n",
       "      <td>8.647458</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>6.658211</td>\n",
       "      <td>5.101538</td>\n",
       "      <td>5.101538</td>\n",
       "      <td>14.394351</td>\n",
       "      <td>6.080373</td>\n",
       "      <td>...</td>\n",
       "      <td>7.872418</td>\n",
       "      <td>9.382984</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>6.658211</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>PM1</td>\n",
       "      <td>2008-01-21</td>\n",
       "      <td>BK</td>\n",
       "      <td>35</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PM1:20071217</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>9.300649</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.249333</td>\n",
       "      <td>5.249333</td>\n",
       "      <td>5.249333</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>13.803455</td>\n",
       "      <td>8.215758</td>\n",
       "      <td>...</td>\n",
       "      <td>7.220602</td>\n",
       "      <td>8.023806</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>5.249333</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>PM1</td>\n",
       "      <td>2007-12-17</td>\n",
       "      <td>BK</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 41 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       sampleID  g__Akkermansia  g__Alistipes  g__Anaerofilum  \\\n",
       "0  PM1:20080108        5.198270      8.861973             0.0   \n",
       "1  PM1:20080114        8.163470      8.685500             0.0   \n",
       "2  PM1:20071211        8.536715      6.230243             0.0   \n",
       "3  PM1:20080121        6.658211      8.647458             0.0   \n",
       "4  PM1:20071217        0.000000      9.300649             0.0   \n",
       "\n",
       "   g__Anaerofustis  g__Anaerostipes  g__Anaerotruncus  g__Anaerovorax  \\\n",
       "0         0.000000         6.178487          6.178487        5.198270   \n",
       "1         0.000000         5.198270          6.496426        0.000000   \n",
       "2         0.000000         6.230243          0.000000        0.000000   \n",
       "3         0.000000         6.658211          5.101538        5.101538   \n",
       "4         5.249333         5.249333          5.249333        0.000000   \n",
       "\n",
       "   g__Bacteroides  g__Bilophila  ...  g__RuminococcaceaeIncertaeSedis  \\\n",
       "0       13.537929      0.000000  ...                         7.488414   \n",
       "1       14.011826      6.178487  ...                         6.756795   \n",
       "2       12.998976      0.000000  ...                         7.220602   \n",
       "3       14.394351      6.080373  ...                         7.872418   \n",
       "4       13.803455      8.215758  ...                         7.220602   \n",
       "\n",
       "   g__Ruminococcus  g__Subdoligranulum  g__Sutterella  g__Turicibacter  \\\n",
       "0         7.750109            0.000000       0.000000         0.000000   \n",
       "1         9.408481            0.000000       4.237039         0.000000   \n",
       "2         0.000000            6.230243       5.249333         5.249333   \n",
       "3         9.382984            0.000000       6.658211         0.000000   \n",
       "4         8.023806            0.000000       5.249333         0.000000   \n",
       "\n",
       "   mouseID        date  diet  relativeTime  status  \n",
       "0      PM1  2008-01-08    BK            22       0  \n",
       "1      PM1  2008-01-14    BK            28       0  \n",
       "2      PM1  2007-12-11    BK             0       0  \n",
       "3      PM1  2008-01-21    BK            35       0  \n",
       "4      PM1  2007-12-17    BK             6       0  \n",
       "\n",
       "[5 rows x 41 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = df_features.merge(df_metadata, left_index=True, right_index=True)\n",
    "df = df.reset_index()\n",
    "df = df.rename(mapper={\"index\":\"sampleID\"}, axis=1)\n",
    "print(df.shape)\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "#df.to_csv(\"INPUT_FILES/mousedata.xls\", sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "35"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bacteria_names = list(df.columns[df.columns.str.contains(\"g__\")])\n",
    "len(bacteria_names)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python (microbiome)",
   "language": "python",
   "name": "microbiome"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
